posts/physical-backup-restore.md

```diff
@@ -0,0 +1,457 @@
+
+To properly test the method, We ended up writing a Bulk Backup and Restoration Tool in FC for finding out maximum possible bugs in this process. 
+
+Everyday we take 40~50 backups and restore the data multiple times in a dummy site.
+
+![](https://frappe.io/files/bulk-restore-tool.png)
+
+
+#### Missing \*.cfg files in Physical Backup
+
+We started seeing some failures where \*.cfg files are missing - preventing InnoDB table imports. This was confusing since AWS EBS Snapshots are supposed to be time consistent.
+
+After some debugging, we found the root cause :
+
+- MariaDB calls `fdatasync` after disk writes to flush data but it doesn't flush the metadata.
+- While the database wrote the \*.cfg files, their metadata remained in memory buffers.
+- Result: Snapshots captured 0-byte \*.cfg files
+
+**Solution -** To mitigate this issue, we did `fsync` for the required `cfg` files before taking the snapshot.
+
+```python
+# Force metadata flush using fsync
+# Ref: https://docs.python.org/3/library/os.html#os.fsync
+with open(file_path, "rb", buffering=0) as f:
+	os.fsync(f.fileno())
+```
+
+**Added Safeguards -** 
+
+1. During Backups:
+	- Store exact file sizes for all files
+	- Calculate checksums for all \*.cfg files
+2. During Restoration:
+	- Verify file sizes first
+	- Verify checksum of \*.cfg files
+	- Abort Restoration if any inconsistencies are detected
+
+#### Handle Broken MyISAM Tables
+
+MyISAM tables are prone to corruption. They corrupt easily but fortunately are just as easy to fix! 😅 
+
+**Common reasons of corruption -**
+
+1. Improperly closed database connections
+2. Unexpected database server crashes
+3. Hardware failure
+
+In most of the cases,
+
+- The index file (**\*MYI**) got corrupted.
+- The data file (**\*MYD**) stays safe.
+
+
+MyISAM comes with a utility called [`myisamchk`](https://mariadb.com/kb/en/myisamchk/) to check and fix corruption and you don't need to run database to do the fixtures.
+
+Just run,
+```
+myisamchk -r /var/lib/mysql/_c8383gdu339jd8/__global_search
+```
+
+It can solve 99% of problem. So, before attempting restoration we check for corruption in MyISAM table files and fix corruption.
+
+If that fails, we will not attempt restoration.
+
+After restoration also, we found some instances where MyISAM Table was marked as corrupted.
+
+![](https://frappe.io/files/myisam-corruption.png)
+
+In this MyISAM restoration procedure, we just copy files. That doesn't update the index position and some metadatas.
+
+So after restoration we check for corruption using
+
+```
+CHECK TABLE <table-name>
+```
+
+If we find any error, we will repair that using
+```
+REPAIR TABLE <table-name>
+```
+
+> We are using `CHECK TABLE` instead of `myisamchk` in post-restoration phase because it's unsafe to use `myismchk` tool while database is running.
+
+#### The FULLTEXT Index Restoration Challenge
```

Comment:
I'd suggest picking the top 3 problems and only discussing them. I was gonna say this during your talk too, the lazy loading part is so interesting that going from that to minor index corrupt feels weird while reading.

Reply:
Reason for suggestion: You're trying to tell a story here, not a step-by-step prescription for how to implement it :)